polygeist/lib/polygeist/Passes/ConvertPolygeistToLLVM.cpp

```diff
@@ -45,16 +49,477 @@
 using namespace mlir;
 using namespace polygeist;
 
-mlir::LLVM::LLVMFuncOp GetOrCreateMallocFunction(ModuleOp module);
-mlir::LLVM::LLVMFuncOp GetOrCreateFreeFunction(ModuleOp module);
+/// Returns the LLVM type of the global variable given the memref type `type`.
+static Type convertGlobalMemrefTypeToLLVM(MemRefType type,
+                                          TypeConverter &typeConverter) {
+  // LLVM type for a global memref will be a multi-dimension array. For
+  // declarations or uninitialized global memrefs, we can potentially flatten
+  // this to a 1D array. However, for memref.global's with an initial value,
+  // we do not intend to flatten the ElementsAttribute when going from std ->
+  // LLVM dialect, so the LLVM type needs to me a multi-dimension array.
+  const auto shape = type.getShape();
+  const auto convElemTy = typeConverter.convertType(type.getElementType());
+  if (!convElemTy)
+    return {};
+  // Shape has the outermost dim at index 0, so need to walk it backwards
+  return std::accumulate(
+      shape.rbegin(), shape.rend(), convElemTy,
+      [](auto ty, auto dim) { return LLVM::LLVMArrayType::get(ty, dim); });
+}
+
+/// States whether a signature should be converted
+///
+/// A signature should be converted if it contains at least a memref type and
+/// all of the memref types in it can be lowered to a bare ptr.
+static bool shouldConvertSignature(TypeRange types) {
+  bool hasMT{false};
+  for (auto type : types) {
+    if (auto mt = type.dyn_cast<MemRefType>()) {
+      if (!canBeLoweredToBarePtr(mt))
+        return false;
+      hasMT = true;
+    }
+  }
+  return hasMT;
+}
```

Comment:
We're being conservative here for now. Lower signatures with at least one memref for which all present memrefs can be converted to the bare ptr convention.